### PR TITLE
Add core::ops::Range*::contains() as per rust-lang/rust#32311

### PR DESCRIPTION
**Pull request by nodakai**

### CODE REVIEW

**Comment by nodakai** [comment]
I wish to have

```rust
impl<Idx: PartialOrd<Idx>> RangeFull {
  fn contains(&self, _: Idx) -> bool { true }
}
```

so that

```
(..).contains(42) // true
(..).contains("forty-two".to_string()) // true
```

but it seems like it's not allowed:

```
error: the type parameter `Idx` is not constrained by the impl trait, self type, or predicates [E0207]
```

**Comment by alexcrichton** [comment]
Travis failures look non-spurious:

```
---- ops::RangeInclusive_0 stdout ----
        <anon>:5:16: 5:19 error: unresolved name `arr` [E0425]
<anon>:5     assert_eq!(arr[ ...2], [0,1,2  ]);
                        ^~~
<anon>:5:5: 5:39 note: in this expansion of assert_eq! (defined in <std macros>)
<std macros>:5:8: 5:18 error: the type of this value must be known in this context
<std macros>:5 if ! ( * left_val == * right_val ) {
                      ^~~~~~~~~~
<anon>:5:5: 5:39 note: in this expansion of assert_eq! (defined in <std macros>)
<std macros>:5:22: 5:33 error: the type of this value must be known in this context
<std macros>:5 if ! ( * left_val == * right_val ) {
                                    ^~~~~~~~~~~
<anon>:5:5: 5:39 note: in this expansion of assert_eq! (defined in <std macros>)
error: aborting due to previous error(s)
thread 'ops::RangeInclusive_0' panicked at 'Box<Any>', src/librustc/session/mod.rs:153
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- ops::RangeToInclusive_0 stdout ----
        <anon>:4:5: 4:13 error: unresolved name `Examples` [E0425]
<anon>:4     Examples
             ^~~~~~~~
error: aborting due to previous error(s)
thread 'ops::RangeToInclusive_0' panicked at 'Box<Any>', src/librustc/session/mod.rs:153

---- ops::RangeToInclusive_1 stdout ----
        <anon>:5:21: 5:25 error: inclusive range syntax is experimental (see issue #28237)
<anon>:5     assert_eq!(arr[ ...2], [0,1,2  ]);  // RangeToInclusive
                             ^~~~
<anon>:5:5: 5:39 note: in this expansion of assert_eq! (defined in <std macros>)
<anon>:5:21: 5:25 help: add #![feature(inclusive_range_syntax)] to the crate attributes to enable
<anon>:6:20: 6:25 error: inclusive range syntax is experimental (see issue #28237)
<anon>:6     assert_eq!(arr[1...2], [  1,2  ]);
                            ^~~~~
<anon>:6:5: 6:39 note: in this expansion of assert_eq! (defined in <std macros>)
<anon>:6:20: 6:25 help: add #![feature(inclusive_range_syntax)] to the crate attributes to enable
error: aborting due to previous error(s)
thread 'ops::RangeToInclusive_1' panicked at 'Box<Any>', src/librustc/session/mod.rs:153


failures:
    ops::RangeInclusive_0
    ops::RangeToInclusive_0
    ops::RangeToInclusive_1
```

Also cc @steveklabnik, quite extensive docs! Not sure if they're idiomatic though?

File: src/libcore/ops.rs [hunk 1]
```diff
@@ -1513,7 +1628,45 @@ impl<Idx: fmt::Debug> fmt::Debug for RangeTo<Idx> {
...
+/// Use `start...end` (three dots) for its shorthand.
+///
+/// See the [`contains()`](#method.contains) method for its characterization.
+///
+/// # Examples
+///
+/// ```
+/// #![feature(inclusive_range,inclusive_range_syntax,iter_arith)]
+/// fn main() {
+///     assert_eq!(arr[ ...2], [0,1,2  ]);
```
Comment (durka) [inline src/libcore/ops.rs]: copy/paste error?

File: src/libcore/ops.rs [hunk 2]
```diff
@@ -1576,7 +1729,53 @@ impl<Idx: PartialOrd + One + Sub<Output=Idx>> From<Range<Idx>> for RangeInclusiv
...
+}
+
+/// An inclusive range which is only bounded above: { x | x <= end }.
+/// Use `...end` (three dots) for its shorthand.
+///
+/// See the [`contains()`](#method.contains) method for its characterization.
+///
+/// It cannot serve as an iterator because it doesn't have a starting point.
+///
+/// ```
```
Comment (durka) [inline src/libcore/ops.rs]: spurious backticks

**Comment by nodakai** [comment]
Oops, I accidentally rewound my working branch

- https://travis-ci.org/rust-lang/rust/builds/117442252
- https://s3.amazonaws.com/archive.travis-ci.org/jobs/117442253/log.txt

File: src/libcore/ops.rs [hunk 3]
```diff
@@ -1458,7 +1475,26 @@ impl fmt::Debug for RangeFull {
     }
 }
 
-/// A (half-open) range which is bounded at both ends.
+/// A (half-open) range which is bounded at both ends: { x | start <= x < end }.
+/// Use `stard..end` (two dots) for its shorthand.
```
Comment (durka) [inline src/libcore/ops.rs]: typo

File: src/libcore/ops.rs [hunk 4]
```diff
@@ -1477,7 +1513,50 @@ impl<Idx: fmt::Debug> fmt::Debug for Range<Idx> {
...
+    /// ```
+    #[unstable(feature = "range_contains",
+               reason = "recently added as per RFC", issue = "32311")]
+    pub fn contains(&self, item: Idx) -> bool {
+        (self.start <= item) && (item < self.end)
+    }
+}
+
+/// A range which is only bounded below: { x | start <= x }.
+/// Use `stard..` for its shorthand.
```
Comment (durka) [inline src/libcore/ops.rs]: typo

**Comment by nodakai** [comment]
Fixed typo
Changed #[unstable] to `impl`-wide; not much difference when each `impl` has a single method.

**Comment by durka** [comment]
Wait does `#[unstable]` on an impl actually do anything?

**Comment by petrochenkov** [comment]
@durka 
>Wait does #[unstable] on an impl actually do anything?

It passes the unstability to all the child items.

**Comment by durka** [comment]
OK, I remembered something about trait impls being insta-stable. But inherent impls are different I guess.

**Comment by nodakai** [comment]
The outputs look like this:
http://pauli.matrix.jp/doc-a21c5f267a722b6c80790a559bf272942b946cd4/std/ops/struct.Range.html

**Pull request merged**